### PR TITLE
Adding Role Management with CanCanCan Authorization

### PR DESCRIPTION
**Pull request by henask12**
### Description

This pull request introduces role management to the user system and implements CanCanCan for fine-grained authorization control. The project requirements addressed in this pull request include:

- [x] Installed CanCanCan to enhance user authorization capabilities.
- [x] Added a `role` column to the users table via a migration.
- [x] Implemented authorization rules for post and comment deletion.
- [x] Added a "Delete" button to the view for authorized users to manage their content.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -38,6 +38,16 @@ def deletelike
...
+  def destroy
+    @post = Post.find(params[:id])
+    Comment.where(post_id: @post).delete_all if Comment.where(post_id: @post).any?
+    user = User.find(params[:user_id])
+    return unless @post.destroy
+
+    @post.update_post_counter
+    redirect_to user_path(user)
+  end
+
```
Comment (topeogunleye) [inline app/controllers/posts_controller.rb]: Hello and splendid job with your efforts on the `PostsController`! 🎉 

Your diligence in crafting the methods and ensuring functionalities are properly handled is evident and commendable. While you've done remarkably well in managing CRUD operations, let's sharpen our focus on aligning every feature with the project’s requirements, especially concerning `CanCanCan` usage for authorization.

#### **Integrating CanCanCan for Authorization:**
- **Project Requirement:** Introduce CanCanCan for Post authorization by utilizing `load_and_authorize_resource` at the commencement of the controller.
- **Current Implementation:** The automatic loading and authorization checks provided by `CanCanCan` seem to be absent in your controller actions.
  
  Suggested Modification:
  ```ruby
  class PostsController < ApplicationController
    load_and_authorize_resource
    
    ...
  end
  ```
  Integrating `load_and_authorize_resource` will automatically manage resource loading and authorization checks across all controller actions, fostering a DRY (Don’t Repeat Yourself) coding environment.

### Potential Refinement in Code Implementation:
```ruby
class PostsController < ApplicationController
  load_and_authorize_resource

  ...
  
  def destroy
    user = User.find(params[:user_id])
    @post.destroy if @post.present?
    redirect_to user_path(user)
  end
  
  private
  
  def posts_params
    params.require(:post).permit(:title, :text)
  end
end
```

### Concluding Thoughts:
Your progression is remarkable and the logic embedded within your controller is robust! Harnessing the full potential of available tools and methodologies (like CanCanCan) will bolster your code’s maintainability and readability. 🚀 

Keep the coding flames burning brightly, and should you have any inquiries or seek additional clarity on the feedback provided, feel free to reach out.

Keep coding brilliantly! 🧑‍💻🚀

Warm Regards,
Temitope

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @henask12, @francksefu,

Good job so far!, Hello and thank you for your diligent efforts on the project! Your work with CanCanCan and role-based authorizations in Rails is taking a great shape.
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪🌟📚

- Embedding CanCanCan:

Fantastic Job with integrating CanCanCan! Including gem 'cancancan' in your Gemfile is crucial for leveraging its functionalities, and you've nailed this step!
- Constructive Ability Class:

A Round of Applause for developing a structured Ability class! 🎉 The rules you've set for Post and Comment models are clear and well articulated. This demonstration of managing user-specific and admin permissions is crucial and implemented effectively.
- Role Column Implementation:

Spot On with the role column in the users table! 🎯 Ensuring this column is in your migration files underpins the role-based authorization strategy and you've ensured it’s appropriately configured.

> ## Required Changes ♻️
> _Check the comments under the review._
> 
> ## Optional suggestions
> _Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your
code better._

You can also consider:
[OPTIONAL]

### Areas for  Improvement: 

You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (henask12) [reply app/controllers/posts_controller.rb]: Resolved Thank you @topeogunleye

Comment (topeogunleye) [reply app/controllers/posts_controller.rb]: @henask12, you're welcome.

**Review by TuyishimireEric — APPROVED** [review]
# Status: APPROVED 🟢  
### Hi  @henask12  👋 👋 

#### I'm @TuyishimireEric, Your code reviewer,
Good job on implementing the project requirements! 🥇 
done

Your project is complete!  There is nothing else to say other than... it's time to merge it 🦾  
Congratulations! 🎉


## Highlight 👍 
✅ Made suggested changes 👍 

### Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/enus/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**